Labels & label-based filtering

tycebrown opened the pull request:
Questions/Comments/Considerations:
- At the current moment, each `ListPreviewWidget` does NOT display the labels of its list. Should this be changed? Maybe add a little 'expand labels' button?
- For filtering, I added a button in the app bar which pulls up a `FilterDialog`. Filtering results based on labels kinda works like the resources filter does for [Bible Gateway](https://www.biblegateway.com/passage/search=proverbs+3%3A5&version=NASB) (scroll down, see to the right); a `List` of [`FilterChip`](https://api.flutter.dev/flutter/material/FilterChip-class.html)s containing different labels is displayed in a `Wrap` widget. Only lists containing at least one of the selected labels are displayed. Should this be changed, perhaps to work more like [google keep](https://keep.google.com/u/0/)?
- How should the GUI be cleaned up?
- Any way to clean the code up?

Closes #19

tycebrown committed:
raw code for displaying, adding, and removing labels in list settings dialog

tycebrown committed:
Got it cleaned up, added filtering

tycebrown committed:
Made list settings dialog display of labels a little prettier

tycebrown committed:
Merge branch 'main' into 19-add-labels-or-tags

tycebrown committed:
Renaming. Conflicts resolved, but still some bugs to fix.

tycebrown committed:
Appears to be working. Needs cleaning before can be reviewed

tycebrown committed:
Added the  handling of deleted labels

tycebrown commented:
A few more points:
- After resolving the conflicts with the main branch, I had to fix some issues. One situation I had to code for was the following: the user selects a label in the filter dialog. Then, he deletes that label from every list that has it. Now, that label is still stored as a selected label. However, it is no longer the label of any list and, when the labels from all the lists are gathered for filtering, that label will not be among them. So, in the original code, the app would be silently filtering based on a label, which may give unexpected results to the user if they were to add that label to another list or remove all visible label filters in the filter dialog (because the app would show no results, since it would still be filtering based on the deleted label. 

  Please let me know if I have not explained this clearly.

  I decided to display the deleted labels with a different style at the end of the filter dialog (as long as they are selected; if they are unselected, they go away since they are only stored in the set of currently selected labels). Have I handled this well? Is the code clean? What would have been a better thing to do?
- Any spaghetti code? Any better designs/structures?

tycebrown committed:
updated the _deletedLabels getter's comment

Pertempto commented on lib/model/list_model.dart:
```diff
@@ -67,6 +72,20 @@ class ListModel {
     }
   }
+
+  // void addLabel(String label) {
```
Why are these commented out?

tycebrown replied:
I'm sorry, I should have removed them. I ended up not needing them. Will remove.

Pertempto commented:
> At the current moment, each ListPreviewWidget does NOT display the labels of its list. Should this be changed? Maybe add a little 'expand labels' button?

Why not just display them on the right side of each `ListPreviewWidget`? If we can't easily see the labels on the home page, this feature is much less useful.

Pertempto commented:
To your question about deleted labels....

When a label is deleted, it should be deleted everywhere. It should be unselected from the filter dialog. If it was the only label selected in the filter, then the filter should go back to the "All" option.

Pertempto commented on lib/view/add_label_dialog.dart:
```diff
@@ -0,0 +1,57 @@
+                onChanged: (value) => setState(() => searchQuery = value),
+                onSubmitted: (newLabel) => Navigator.pop(context, newLabel))),
+        SizedBox(
+          height: 200,
```
This doesn't need to be so tall. It could also have some top padding to give space between the first label `ListTile` and the text input.

Pertempto commented on lib/view/add_label_dialog.dart:
```diff
@@ -0,0 +1,57 @@
+  Widget build(BuildContext context) {
+    return SimpleDialog(
+      children: [
+        ListTile(
```
Why is the `TextField` wrapped in a `ListTile`?

tycebrown replied:
I was trying to get the `TextField` lined up with the `ListTile`s containing the labels from other lists. I should not have done that this way. Will replace the `ListTile` with `Padding`.

Pertempto commented on lib/view/list_settings_dialog.dart:
```diff
@@ -29,10 +33,47 @@ class _ListSettingsDialogState extends State<ListSettingsDialog> {
   Widget build(BuildContext context) {
     return AlertDialog(
       title: const Text('Enter New List Title', textAlign: TextAlign.center),
```
We really could improve this title. Something like "List Settings"

tycebrown committed:
Removed commented-out addLabel and removeLabel. Also changed _getFilteredData to use listModel.hasLabel instead of listModel.labels.contains

tycebrown committed:
re titled the list settings dialog

tycebrown committed:
Removed ListTile around TextField and replaced with Padding,Reduced the height of the add label dialog's ListView of labels

Pertempto commented:
Also, please resolve the merge conflicts.

tycebrown committed:
Fixed the deleted labels problem

tycebrown committed:
Merge branch 'main' into 19-add-labels-or-tags

tycebrown committed:
Added labels to right of list preview widget

tycebrown commented:
Conflicts resolved! Now, labels are displayed on the right of the `ListPreviewWidgets`. I don't know how this shows up on mobile though.

Pertempto commented:
> Conflicts resolved! Now, labels are displayed on the right of the `ListPreviewWidgets`. I don't know how this shows up on mobile though.

Are you ready for another review? Please click the "request review" button so I know for sure.

Pertempto commented on lib/view/home_page.dart:
```diff
@@ -61,6 +61,7 @@ class _HomePageState extends State<HomePage> {
       builder: (context, snapshot) {
         if (snapshot.hasData) {
           allLabels = _getAllLabels(of: snapshot.data!);
+          _removeSelectedDeletedLabels();
```
I would just put the contents of this method here, because it is only 2 lines and there is no need to add another function that is only called once.

Pertempto commented on lib/view/list_preview_widget.dart:
```diff
@@ -37,8 +37,26 @@ class _ListPreviewWidgetState extends State<ListPreviewWidget> {
       onLongPress: _showOptionsModalSheet,
-      title: Text(widget.listModel.title,
-          style: Theme.of(context).textTheme.titleLarge),
+      title: Row(
```
No need to create a `Row` here in the `Title` slot. Just put your `Wrap` in the `trailing` slot of the `ListTile`

tycebrown replied:
Well, I was having some trouble with putting the `Wrap` in the `trailing` slot. The labels would not wrap until they hit the far edge of the `ListTile`, forcing the title to softwrap like mad.

here are some screenshots:



![titlewrapproblem](https://github.com/Pertempto/Lists/assets/91034018/65dbc95c-b75f-41f7-865c-5df2fe97cb93)



![evenmoretitlewrapping](https://github.com/Pertempto/Lists/assets/91034018/bfaf7b2a-7a4c-400a-8869-ea3c5eb66a6d)

Pertempto replied:
hahaha whoa that is a problem.

Pertempto replied:
In this case, we are running up against the limits of what `ListTile` can do. I pushed a commit that replaces it with a custom `InkWell`. I used a `Container` to force the labels to stay in a limited width, so it looks better on mobile. Make any adjustments that you need.

Pertempto reviewed (changes requested):
I went ahead and reviewed your changes. It is looking better!

Feature idea: 
Users could tap on one of the labels shown on the home page, and it would be added to the filter. Then they don't have to open the filter dialog! Less taps! 😁    
How difficult would this be? If it's complicated, we can create a separate issue for it. If it's simple, go ahead and add it here.

tycebrown committed:
inlined call to _removeSelectedDeletedLabels

tycebrown commented:
> I went ahead and reviewed your changes. It is looking better!
> 
> Feature idea: Users could tap on one of the labels shown on the home page, and it would be added to the filter. Then they don't have to open the filter dialog! Less taps! grin How difficult would this be? If it's complicated, we can create a separate issue for it. If it's simple, go ahead and add it here.

Do you mean the labels to the right of `ListPreviewWidgets`? I think that should be simple to implement.

Pertempto commented:
> > I went ahead and reviewed your changes. It is looking better!
> > Feature idea: Users could tap on one of the labels shown on the home page, and it would be added to the filter. Then they don't have to open the filter dialog! Less taps! grin How difficult would this be? If it's complicated, we can create a separate issue for it. If it's simple, go ahead and add it here.
> 
> Do you mean the labels to the right of `ListPreviewWidgets`? I think that should be simple to implement.

Yeah. I think you can use `ActionChip`

Commit:
Replace ListTile with InkWell

Commit:
Dynamic width.

tycebrown committed:
Merge branch '19-add-labels-or-tags' of https://github.com/Pertempto/Lists into 19-add-labels-or-tags

tycebrown committed:
small tweak: if the list has no labels, then the widget to display them is not created.

tycebrown committed:
Now, the user can select labels from the home page. Additional small tweak

I made selectedLabels a Set, since it made adding and removing individual labels more convenient.

tycebrown commented:
I ended up using `FilterChip` over `ActionChip`, since `FilterChip` can be marked as selected or unselected (so the user sees the selected labels in each `ListPreviewWidget`).

Pertempto reviewed (approved):
This works very nicely! I'm not sure if we even need the `FilterDialog` anymore, because your new method works so well. I like it!
